scrapyd_k8s/object_storage/libcloud_driver.py

```diff
@@ -0,0 +1,79 @@
+import os
+import logging
+
+logger = logging.getLogger(__name__)
+
+from libcloud.storage.types import ObjectError, ContainerDoesNotExistError, ObjectDoesNotExistError, InvalidContainerNameError
+from libcloud.storage.providers import get_driver
+
+class LibcloudObjectStorage:
+
+    def __init__(self, config):
+        self._storage_provider = config.joblogs().get('storage_provider')
+        if self._storage_provider is None:
+            logger.error('Storage provider is not defined in the configuration.')
+            raise ValueError('Storage provider is not defined')
+
+        self._container_name = config.joblogs().get('container_name')
+        if self._container_name is None:
+            logger.error('Container name is not set in the configuration.')
+            raise ValueError('Container name is not set')
+
+        args_envs = config.joblogs_provider_args(self._storage_provider)
+        args = {}
+        for arg, env in args_envs.items():
+            env_value = os.getenv(env)
+            if env_value is None:
+                logger.error(f"Environment variable '{env}' for argument '{arg}' is not set.")
+                raise ValueError(f"Environment variable '{env}' for argument '{arg}' is not set.")
+            args[arg] = env_value
+
+        driver_class = get_driver(self._storage_provider)
+        print("CLOUD CREDENTIALS")
+        print(args)
+        try:
+            self.driver = driver_class(**args)
+            logger.info(f"Initialized driver for storage provider '{self._storage_provider}'.")
+        except Exception as e:
+            logger.exception(f"Failed to initialize driver for storage provider '{self._storage_provider}'.")
+            raise
+
+    def upload_file(self, local_path: str):
+        object_name = os.path.basename(local_path)
+        try:
+            container = self.driver.get_container(container_name=self._container_name)
+            self.driver.upload_object(
+                local_path,
+                container,
+                object_name,
+                extra=None,
+                verify_hash=True,
```

Comment:
Just to be understand: why did you choose to verify the hash here, is that common practice? Does it incur extra costs?

Reply:
It is a good practice because this was we verify that the file has not been corrupted during transmission, if you think it's not critical we can skip it, of course.
It does incur extra cpu costs and for very large files the time can be noticeable.

If we don't care about data integrity here and fast upload is the primary goal, then I'll change it to
```
verify_hash=False
```

Reply:
Thanks for explaning. Let's not enable it by default. It the needs comes up later, we can add it with an extra option (but then we have a whole range of S3-related things we may or may not want to enable, like storage class, acls, perhaps versioning, policies, e.a.).